### DISCUSSION TITLE
Frosting - Manually running tasks ignores IsDependentOn?

### DISCUSSION THREAD

**Discussion by p10tyr** [cake-build/cake#3541] [Extension Q&A] [answered]
I am trying out Frosting because I wanted for a very long time a way to make a user selection menu based on the ... many tasks we have.

So I am not all that clued up the internal API's and capabilities. I figured out I can create my own instance of a Task and all things programatically.. but I wanted to keep the Tasks in a folder with separate files.

I am using the Default task to ..
- Use reflection on current assembly to find everything that implements IFrostingTask (AsyncFrostingTask<BuildContext>)
- Got a menu working using Spectre
- Activate the instance using Instance.Activator and cast it to IFrostingTask 
- Instantiate BuildContext using ICakeContext
- Pass that into the activated class using RunAsync(buildContext);

Sweet . that works great. But it only runs that ONE task and the IsDependentOn is ignored.

I guess this is something the Core.Engine is responsible for.. 

Is there another way I should be doing this?

Thanks for a great piece of software - been using the ps1 script for many, many years at various places. 💯

**Comment by augustoproiete** [comment, marked as answer]
Cześć! It looks like you're missing some components to have the Cake engine run in a similar fashion as a regular Cake/Frosting build. Take a look at the community projects [`Cake.Bridge`](https://github.com/devlead/Cake.Bridge) and/or [`Cake.Console`](https://github.com/pitermarx/Cake.Console) to see how they're doing it.

**Reply by p10tyr** [reply]
Cześć! 😄 
So I really want to stay with Frosting since its got first class support from CAKE. This is good because developers can reference the CAKE website and not have to deal with other variations of it.

So I had a look at the two links you provided and it made me realise... 
- Keep the frosting project as explained in the documentation
- Make a separate console app that finds the tasks with reflection
- On choice.. invoke cake with the provided argument

That is is all I wanted to do really have the ability to make like a splash screen and have everything else box standard.

Thank you for your time! 💯

**Reply by p10tyr** [reply]
Quick update actually. When I was doing my own console app.. I realised that arguments get passed in to Main method.
```public static int Main(string[] args)```

So all I really had to do is put the menu selector before ```return new CakeHost()...```

If anybody passes in arguments ill just skip the menu.. if it comes in with no arguments.. I can mutate them as needed.

I have included the code in case somebody else may find it useful. 

```
 public static int Main(string[] args)
    {
        args = SplashScreen(args);

        return new CakeHost()
            .UseContext<BuildContext>()
            .Run(args);
    }

    public static string[] SplashScreen(string[] args)
    {
        if (args.Length > 0)
            return args;

        var tasks = GetAllTasks();

        var selectPromtp = new SelectionPrompt<string>()
               .Title("[green]Select the task you would like to run[/]")
               .AddChoices(tasks.ToArray());
       
        var selectedTask = AnsiConsole.Prompt<string>(selectPromtp);

        Console.WriteLine($"Selected {selectedTask} ");

        return new string[] { $"-t", selectedTask };
    }

    public static List<string> GetAllTasks()
    {
        return AppDomain.CurrentDomain.GetAssemblies().SelectMany(x => x.GetTypes())
                .Where(x => typeof(IFrostingTask).IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract)
                .Select(x => x.FullName)
                .Select(fullyQualifiedTaskName =>
                {
                    var attribute = (TaskNameAttribute)Attribute
                        .GetCustomAttribute(Type.GetType(fullyQualifiedTaskName), typeof(TaskNameAttribute));
                    return attribute.Name;
                })
                .ToList();
    }
```
